### ISSUE TITLE
X-Pack installing trial license when set to 'No' with v7.3.0

### ISSUE THREAD

**Issue by justin-silveri** [elastic/azure-marketplace#303]
According to the Readme.md, when 'xpackPlugins = No' and 'esVersion >= 6.3', it should install the basic license for Xpack. When we created the cluster with xpack set to no and es version set to 7.3.0, the elasticsearch.yml was set with "xpack.security.enabled: true" and "xpack.license.self_generated.type: trial". 

Looking at the script, it checks for esversion>= 7.1.0 or esversion >= 6.8.0 and <7.0.0 , then sets BASIC_SECURITY=1:
https://github.com/elastic/azure-marketplace/blob/3a9f52ad2c886cd0d1c4d3118f5fbcad3c45d416/src/scripts/elasticsearch-install.sh#L272

Which then triggers the script to install xpack security with trial license, even though INSTALL_XPACK is set to 0:
https://github.com/elastic/azure-marketplace/blob/3a9f52ad2c886cd0d1c4d3118f5fbcad3c45d416/src/scripts/elasticsearch-install.sh#L1057

So I'm not seeing how it's possible to install the basic license. Am I missing something?

**Comment by russcam** [comment]
Thank for opening this, @justin-silveri. This looks like a bug; on line 1057, the trial license should be set only when `INSTALL_XPACK=1`. I'll look to open a PR to fix.

In the meantime, you can revert to a basic license, either by 

- using the Start basic license button in Kibana, under Management, if Kibana is installed:

    ![image](https://user-images.githubusercontent.com/208231/63474241-ca17da00-c4bb-11e9-84c0-862aff3ba327.png)

or 

- [callling the start basic license API](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/start-basic.html):

    ```sh
    POST /_license/start_basic?acknowledge=true
    ```

`acknowledge=true` is passed to acknowledge that reverting to a basic license loses features.

 In both cases, license information is stored in cluster state, and supersedes the setting in elasticsearch.yml. **Please note that using the basic license with security enabled (which is highly recommended) requires _at least_ [configuring TLS for the Elasticsearch Transport layer](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/configuring-tls.html#tls-transport).** Reverting to the basic license should be done _after_ configuring TLS for the transport layer. If you need more information about this, please let me know.

**Comment by russcam** [comment]
> I've tested the script changes on our cluster, and while it no longer sets the license to 'trial', it does not set it to basic. 

Which script changes are you testing, and how are you testing them?

In 6.3.0+, when no `xpack.license.self_generated.type` is configured in elasticsearch.yaml, the default distribution resorts to a basic license.

There's a test to check for this:

I just ran this with the following on the `fix/303` branch

```sh
npm run release -- --esVersion 6.5.4 --test 1d-0m-0c-basic-ext-kp --nodestroy
```

and this is what is in License Management in Kibana:

![image](https://user-images.githubusercontent.com/208231/63730374-62440380-c8ae-11e9-8785-30d7792e9008.png)

![image](https://user-images.githubusercontent.com/208231/63730348-480a2580-c8ae-11e9-8cd7-b5ee5d705b9d.png)

> It also still sets the 'xpack.security.enabled: true' in the elasticsearch.yml config.

This is intentional when the version being deployed is >= 6.8.0 and < 7.0.0, or >= 7.1.0, to always configure Security, irrespective of license applied. There's a test also to check for this, which I've ran on the `fix/303` branch too:

```sh
npm run release -- --esVersion 7.3.0 --test 1d-0m-0c-basic-ext-tls-klp --nodestroy
```

This deploys 

![image](https://user-images.githubusercontent.com/208231/63731163-b3a1c200-c8b1-11e9-9919-af87d2637919.png)

![image](https://user-images.githubusercontent.com/208231/63731139-a1278880-c8b1-11e9-95c2-c8a2a8d3b04c.png)

> Kibana still says it's running on a 'trial' license.

I'm not sure if Kibana caches license details. If a deployment was performed into an existing resource group with an existing deployment, Kibana _may_ need restarting to reflect the license change. If need be, it can be restarted with

```sh
sudo systemctl restart kibana.service
```

**Comment by russcam** [comment]
Reopening; was closed as a result of #304 being merged.

**Comment by russcam** [comment]
@justin-silveri template version 7.3.1 is now released. Would like to give it a try by targeting the `7.3.1` tag?

For example, with the Azure CLI 2.0

```sh
template_base_uri=https://raw.githubusercontent.com/elastic/azure-marketplace
template_version=7.3.1

az group deployment create \
  --resource-group "<name>" \
  --template-uri $template_base_uri/$template_version/src/mainTemplate.json \
  --parameters _artifactsLocation=$template_base_uri/$template_version/src/ \
               esVersion=7.3.1 esClusterName=elasticsearch \
               vmDataDiskCount=1 dataNodesAreMasterEligible=Yes \
               adminUsername=russ adminPassword=Password1234 \
               securityBootstrapPassword=bootstrapPassword123 \
               securityAdminPassword=adminPassword123 \
               securityKibanaPassword=kibanaPassword123 \
               securityLogstashPassword=logstashPassword123 \
               securityBeatsPassword=beatsPassword123 \
               securityApmPassword=apmPassword123 \
               securityRemoteMonitoringPassword=remoteMonitoringPassword123
```

**Comment by russcam** [comment]
> I tested the merged changes to the elasticsearch-install.sh and updated the stack. The elasticsearch.yml now only has 'xpack.security.enabled: true', with no reference to the license type (basic or trial). I believe a stack update already restarts the data nodes as well as kibana, since kibana is unreachable for 1-2 minutes when that resource is updating. When I log on kibana, it still shows it's running on the trial license, which is about to expire.

Sounds to me like you're applying the changes to an already running cluster, which I wouldn't really advise doing, across template versions; if it's just the changes to ` elasticsearch-install.sh`, I think that'll probably be OK in this instance, as it only has the fix for setting license type.

> Where is that referenced, and how do I revert back to the basic license at the server level?

The trial license information will be held in cluster state now. You'll be able to revert this to a basic license as per https://github.com/elastic/azure-marketplace/issues/303#issuecomment-523684202, using the start basic license API or within Kibana.

The integration tests have passed with the changes from #304, and it's now been released to the Marketplace in [7.3.1](https://github.com/elastic/azure-marketplace/releases/tag/7.3.1), so I'm going to close this issue as resolved.
